### ISSUE TITLE
netCDF Split&Swap

### ISSUE THREAD

**Issue by zakjan** [OSGeo/gdal#5497]
## Expected behavior and actual behavior.

GRIB Split&Swap mode (#4526) would be useful for netCDF driver as well. What approach would be suitable to generalize the feature? Copy it to netCDF? Or expose as a separate gdal_translate/gdalwarp flag, so that it is supported by all drivers?

## Steps to reproduce the problem.

Download source file [sst_anomaly.nc.zip](https://github.com/OSGeo/gdal/files/8337193/sst_anomaly.nc.zip) (comes from [CMEMS](https://resources.marine.copernicus.eu/product-detail/SST_GLO_SST_L4_NRT_OBSERVATIONS_010_001/DATA-ACCESS)).

```
$ gdalinfo NETCDF:sst_anomaly.nc:sst_anomaly
...
Size is 1440, 720
Origin = (0.000000000000000,90.000000000000000)
Pixel Size = (0.250000000000000,-0.250000000000000)
...
Corner Coordinates:
Upper Left  (   0.0000000,  90.0000000)
Lower Left  (   0.0000000, -90.0000000)
Upper Right (     360.000,      90.000)
Lower Right (     360.000,     -90.000)
Center      ( 180.0000000,   0.0000000)
```

gdal_translate it to GeoTIFF. Expected result is in -180..180 longitude range, same as GRIB. Actual result is in 0..360 longitude range.

## Operating system

macOS 12.1 M1

## GDAL version and provenance

3.4.2 from brew

**Comment by rouault** [comment]
> Or expose as a separate gdal_translate/gdalwarp flag

That should reasonably fit in gdal_translate architecture (controlled by an option), by using underneath a VRT assembling 2 sources..
gdalwarp -t_srs "+proj=longlat +datum=WGS84 +lon_wrap=0" could perhaps do the job too.

**Comment by zakjan** [comment]
Yes, I use `+lon_wrap` currently, but it requires a separate call to gdalwarp with a non-standard projection string and a temporary file. I like how the new GRIB Split&Swap allows to use a single gdal_translate command. Is there support for this transformation in VRT already?

**Comment by rouault** [comment]
> Is there support for this transformation in VRT already?

no, I gave above the hint how that could possibly be implemented, but work needed

**Comment by mmomtchev** [comment]
This old stackoverflow by @warmerdam contains the right command:
https://gis.stackexchange.com/questions/37790/reprojecting-raster-from-0-360-to-180-180-with-cutting-180-meridian-using-gdalw

```
gdalwarp -s_srs "+proj=longlat +ellps=WGS84" -t_srs WGS84 ~/0_360.tif 180.tif  -wo SOURCE_EXTRA=1000 \
           --config CENTER_LONG 0
```

GRIBv2 was a very particular case where the format itself specified that 0-360° longitudes are to be used

**Comment by zakjan** [comment]
There is my comment as well, CENTER_LONG didn't seem to work well. https://gis.stackexchange.com/questions/37790/reprojecting-raster-from-0-360-to-180-180-with-cutting-180-meridian-using-gdalw#comment665452_223966

```
gdalwarp -s_srs '+proj=longlat +datum=WGS84 +lon_wrap=180' -t_srs '+proj=longlat +datum=WGS84'
```

I get it that if we were to support the feature outside of gdalwarp, it should be a VRT transformation. Thanks, I'll take a look at it.

**Comment by mdsumner** [comment]
You could do 

```bash
gdalwarp -te -180 -90 180 90 vrt:///vsizip//vsicurl/https://github.com/OSGeo/gdal/files/8337193/sst_anomaly.nc.zip/sst_anomaly.nc?a_ullr=-360,90,0,-90  /vsizip//vsicurl/https://github.com/OSGeo/gdal/files/8337193/sst_anomaly.nc.zip/sst_anomaly.nc atlantic.tif

```


If you only do one input, without adding the modified one you only get the east hemisphere, because the warper doesn't wrap (but it could I think). Potentially related to #6582

**Comment by dbaston** [comment]
This is still a fairly tricky thing to do in 2026. I'm thinking it makes sense to implement a generic `gdal raster rotate-longitude` that assembles a VRT under the hood to produce whatever longitude range you want. Would also be useful to have a `gdal vector rotate-longitude` akin to the PostGIS `ST_WrapX`.

**Comment by mdsumner** [comment]
fwiw the warper is already pretty solid on this (-720,360) seems to be the limit

using an R wrapper on the warp api

```R
Sys.setenv(GDAL_NETCDF_ASSUME_LONGLAT="YES")
library(vapour)
library(ximage)
dsn <- "vrt:///vsizip//vsicurl/https://github.com/OSGeo/gdal/files/8337193/sst_anomaly.nc.zip/sst_anomaly.nc?a_ullr=-360,90,0,-90"

## note this uses xmin,xmax,ymin,ymax *extent* as per terra
imwarpfun <- function(dsn, extent = c(-180, 180, -90, 90)) {
  ximage(d <- gdal_raster_data(dsn, target_ext = extent, target_dim = c(360, 0)))
  maps::map(add = T);maps::map('world2', add = T)
  invisible(d)
}
par(mfrow = c(2, 2))
imwarpfun(dsn, c(-360, 0, -90, 90))
imwarpfun(dsn, c(180, 360, -90, 90))
imwarpfun(dsn, c(-130, 230, -90, 90))
imwarpfun(dsn, c(-720, 360, -90, 90))
```

<img width="815" height="647" alt="Image" src="https://github.com/user-attachments/assets/dacb5283-f86c-4784-bd92-c2b78413456b" /> 


(remaining problems I have had were about geolocation arrays with non-standard longitude ranges). Not sure if a warp solution is welcome but just wanted to share that's what I rely on these days.

**Comment by dbaston** [comment]
Yes, it's doable but pretty obscure -- why am I setting `GDAL_NETCDF_ASSUME_LONGLAT="YES"`, opening the file with `vrt://` and overriding the origin point? It also appears to fail for source longitude ranges slightly different from (0, 360) -- for example ERA5 which uses (-0.125, 359.875). A dedicated command should be faster since it would not need to go through the warper, but maybe it's not enough difference to matter.

I'm happy to write the command, just trying to give opportunities to critique the approach before expending the effort...

For example, a disadvantage to an explicit command is that it doesn't provide much of a solution to QGIS users, unlike a "split-and-swap" opening option.

**Comment by rouault** [comment]
> I'm happy to write the command

This feels like a good idea . We don't want to encode that logic in every driver where that can happen.  The issue with geolocation arrays is certainly a complication.
It looks like it could be solved this way: the "virtual" dataset (which is likely to be a VRT) returned by split-and-swap should have a GEOLOCATION metadata with X_DATASET and Y_DATASET being inline VRT doing the split-and-swap on the original X_DATASET/Y_DATASET  .  GCPs if present should also be edited to modify the dfPixel value .  If there is RPC, I believe we can't do anything but discard it with a warning.

**Comment by mdsumner** [comment]
> why am I setting GDAL_NETCDF_ASSUME_LONGLAT="YES", opening the file with vrt:// and overriding the origin point?

those are just augmenting the example here which has no crs set, nothing to do with the operation to target grid from a non-broken file source  It would be needed to set an output crs manually for output from a translate/swap process from such a file. (and sorry I think a_ullr is entirely unnecessary,  though maybe it was needed in 2024 I didn't check). mostly I was just putting out there that it works, I like it because it's general to almost *any target grid* and swapping longitudes is just one special case.  I never liked that raster/terra calls it "rotate", and split-and-swap seems a reasonable name.
